### PR TITLE
Code for RES_NET_ROUTE_ALL.

### PR DESCRIPTION
**Pull request by anjalidk**

### CODE REVIEW

File: net_route.c [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+#include <stdlib.h>
+#include <unistd.h>
+#include <strings.h>
+#include <errno.h>
+#include <string.h>
+#include "resource_impl.h"
+    
+static int count, ecount;
+
+//#define TESTING
```
Comment (hallyn) [inline net_route.c:21]: TESTING doesn't show up in the rest of this file.

File: net_route.c [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+			rt->dst_prefix_len = m->rtm_dst_len;
+		}
+		if (m->rtm_family == AF_INET6) {
+			memcpy(rt->dest, RTA_DATA(at[RTA_DST]), MAX_BYTES_IPV6);
+			rt->dst_prefix_len = m->rtm_dst_len;
+		}
+	}
+	if (at[RTA_SRC]) {
+		if (m->rtm_family == AF_INET) {
+			printf("RTA_SRC for AF_INET!!\n");
```
Comment (hallyn) [inline net_route.c:98]: Did you want this block under PRINTLOGS?

File: net_route.c [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+}
+
+static int get_route_len(int net_sock)
+{
+	int recvl, err;
+	struct msghdr msg;
+	struct iovec iov;
+	struct sockaddr_nl nladdr;
+
+	msg.msg_name = &nladdr;
```
Comment (hallyn) [inline net_route.c:140]: Did you want to memset msg and iov like you do elsewhere?

File: net_route.c [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+	struct iovec iov;
+	struct msghdr msg;
+	struct nlmsghdr *r;
+	struct nlmsgerr *nlmsg_err;
+	struct rtmsg *rt;
+	struct rtattr *at[RTA_MAX + 1];
+	struct rt_info *routes = NULL, *iroutes = NULL;
+	struct sockaddr_nl nladdr;
+
+	msg.msg_name = &nladdr;
```
Comment (hallyn) [inline net_route.c:183]: memset msg?

File: net_route.c [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+				free(routes);
+			return -1;
+		}
+		char *buf = (char *)malloc(len);
+		if (!buf) {
+			if (routes)
+				free(routes);
+			return -1;
+		}
+		iov.iov_base = buf;
```
Comment (hallyn) [inline net_route.c:201]: memset iov for good measure?

File: net_route.c [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+#ifdef PRINTLOGS
+			printf("len after sub %lu is %d\n",sizeof(*rt), len);
+#endif
+			parse_attr(at, RTM_RTA(rt), len);
+			if (rt->rtm_family == AF_INET || rt->rtm_family == AF_INET6) {
+				get_attr(at, iroutes, rt);
+				iroutes++;
+				rtind++;
+			} else {
+				printf("Family is %d\n",rt->rtm_family);
```
Comment (hallyn) [inline net_route.c:277]: Did you want this printf under PRINTLOGS?

File: resource.h [hunk 7]
```diff
@@ -113,6 +113,10 @@ typedef struct res_blk {
 #define RES_NET_ALLIFSTAT		2050
 #define NET_MAX				2051
 
+#define ROUTE_MIN                       2000
```
Comment (hallyn) [inline resource.h:116]: Why not move this above NET_MIN (2048) ?

**Review by hallyn — COMMENTED** [review]
Just a few nitpicks, I don't think there was anything serious.

You might want to switch from '// #define PRINTLOGS' to a "-DPRINTLOGS" which is conditionally enabled in Makefile.

**Comment by anjalidk** [comment]
Thanks Serge, see below:

From: Serge Hallyn ***@***.***>
Date: Saturday, July 8, 2023 at 9:50 AM
To: lxc/libresource ***@***.***>
Cc: Anjali Kulkarni ***@***.***>, Author ***@***.***>
Subject: Re: [lxc/libresource] Code for RES_NET_ROUTE_ALL. (PR #22)

@hallyn commented on this pull request.

Just a few nitpicks, I don't think there was anything serious.

You might want to switch from '// #define PRINTLOGS' to a "-DPRINTLOGS" which is conditionally enabled in Makefile.

ANJALI> Will do.

________________________________

In net_route.c<https://urldefense.com/v3/__https:/github.com/lxc/libresource/pull/22*discussion_r1257309760__;Iw!!ACWV5N9M2RV99hQ!LeZ-PE7I0x3Iv-_s9xznDWtPOzIMUrfEjxIJKOGYh5fwP52jsLmX43K1nSg9z55unjVklZr3uqqOmMhvwYJ-ISAlXE7kmyE$>:

> +#include <linux/rtnetlink.h>

+#include <linux/if_link.h>

+#include <linux/if_arp.h>

+#include <linux/types.h>

+

+#include <stdio.h>

+#include <stdlib.h>

+#include <unistd.h>

+#include <strings.h>

+#include <errno.h>

+#include <string.h>

+#include "resource_impl.h"

+

+static int count, ecount;

+

+//#define TESTING

TESTING doesn't show up in the rest of this file.

ANJALI> Will remove

________________________________

In net_route.c<https://urldefense.com/v3/__https:/github.com/lxc/libresource/pull/22*discussion_r1257309841__;Iw!!ACWV5N9M2RV99hQ!LeZ-PE7I0x3Iv-_s9xznDWtPOzIMUrfEjxIJKOGYh5fwP52jsLmX43K1nSg9z55unjVklZr3uqqOmMhvwYJ-ISAlUoD-ZtA$>:

> +            rt->table = *(int *)RTA_DATA(at[RTA_TABLE]);

+       if (at[RTA_OIF])

+              rt->index = *(int *)RTA_DATA(at[RTA_OIF]);

+       if (at[RTA_DST]) {

+              if (m->rtm_family == AF_INET) {

+                      memcpy(rt->dest, RTA_DATA(at[RTA_DST]), MAX_BYTES_IPV4);

+                      rt->dst_prefix_len = m->rtm_dst_len;

+              }

+              if (m->rtm_family == AF_INET6) {

+                      memcpy(rt->dest, RTA_DATA(at[RTA_DST]), MAX_BYTES_IPV6);

+                      rt->dst_prefix_len = m->rtm_dst_len;

+              }

+       }

+       if (at[RTA_SRC]) {

+              if (m->rtm_family == AF_INET) {

+                      printf("RTA_SRC for AF_INET!!\n");

Did you want this block under PRINTLOGS?

ANJALI> Will move it.

________________________________

In net_route.c<https://urldefense.com/v3/__https:/github.com/lxc/libresource/pull/22*discussion_r1257309877__;Iw!!ACWV5N9M2RV99hQ!LeZ-PE7I0x3Iv-_s9xznDWtPOzIMUrfEjxIJKOGYh5fwP52jsLmX43K1nSg9z55unjVklZr3uqqOmMhvwYJ-ISAl952Hr9Q$>:

> +     while (RTA_OK(rta, len)) {

+              if (rta->rta_type <= RTA_MAX)

+                      at[rta->rta_type] = rta;

+              rta = RTA_NEXT(rta, len);

+       }

+       return 0;

+}

+

+static int get_route_len(int net_sock)

+{

+       int recvl, err;

+       struct msghdr msg;

+       struct iovec iov;

+       struct sockaddr_nl nladdr;

+

+       msg.msg_name = &nladdr;

Did you want to memset msg and iov like you do elsewhere?

ANJALI> Will do.

________________________________

In net_route.c<https://urldefense.com/v3/__https:/github.com/lxc/libresource/pull/22*discussion_r1257309907__;Iw!!ACWV5N9M2RV99hQ!LeZ-PE7I0x3Iv-_s9xznDWtPOzIMUrfEjxIJKOGYh5fwP52jsLmX43K1nSg9z55unjVklZr3uqqOmMhvwYJ-ISAlhNzFbbQ$>:

> +}

+

+static int handle_route_resp(int net_sock, void **out)

+{

+       int err, recvl, len, max_routes, rt_size;

+       int rtind = 0, total_rt_size = 0;

+       struct iovec iov;

+       struct msghdr msg;

+       struct nlmsghdr *r;

+       struct nlmsgerr *nlmsg_err;

+       struct rtmsg *rt;

+       struct rtattr *at[RTA_MAX + 1];

+       struct rt_info *routes = NULL, *iroutes = NULL;

+       struct sockaddr_nl nladdr;

+

+       msg.msg_name = &nladdr;

memset msg?

ANJALI> Will do.

________________________________

In net_route.c<https://urldefense.com/v3/__https:/github.com/lxc/libresource/pull/22*discussion_r1257309965__;Iw!!ACWV5N9M2RV99hQ!LeZ-PE7I0x3Iv-_s9xznDWtPOzIMUrfEjxIJKOGYh5fwP52jsLmX43K1nSg9z55unjVklZr3uqqOmMhvwYJ-ISAlInwec1Q$>:

> +     msg.msg_iovlen = 1;

+

+       while(1) {

+              len = get_route_len(net_sock);

+              if (len == -1) {

+                      if (routes)

+                              free(routes);

+                      return -1;

+              }

+              char *buf = (char *)malloc(len);

+              if (!buf) {

+                      if (routes)

+                              free(routes);

+                      return -1;

+              }

+              iov.iov_base = buf;

memset iov for good measure?

ANJALi> Yes.

________________________________

In net_route.c<https://urldefense.com/v3/__https:/github.com/lxc/libresource/pull/22*discussion_r1257310486__;Iw!!ACWV5N9M2RV99hQ!LeZ-PE7I0x3Iv-_s9xznDWtPOzIMUrfEjxIJKOGYh5fwP52jsLmX43K1nSg9z55unjVklZr3uqqOmMhvwYJ-ISAlSXSQrZ0$>:

> +                            eprintf("nlmsg_len incorrect");

+                              errno = EINVAL;

+                              free(buf);

+                              free(routes);

+                              return -1;

+                      }

+#ifdef PRINTLOGS

+                      printf("len after sub %lu is %d\n",sizeof(*rt), len);

+#endif

+                      parse_attr(at, RTM_RTA(rt), len);

+                      if (rt->rtm_family == AF_INET || rt->rtm_family == AF_INET6) {

+                              get_attr(at, iroutes, rt);

+                              iroutes++;

+                              rtind++;

+                      } else {

+                              printf("Family is %d\n",rt->rtm_family);

Did you want this printf under PRINTLOGS?

ANJALI> Will move it.

________________________________

In resource.h<https://urldefense.com/v3/__https:/github.com/lxc/libresource/pull/22*discussion_r1257310597__;Iw!!ACWV5N9M2RV99hQ!LeZ-PE7I0x3Iv-_s9xznDWtPOzIMUrfEjxIJKOGYh5fwP52jsLmX43K1nSg9z55unjVklZr3uqqOmMhvwYJ-ISAlJDWd62k$>:

> @@ -113,6 +113,10 @@ typedef struct res_blk {

 #define RES_NET_ALLIFSTAT            2050

 #define NET_MAX                              2051



+#define ROUTE_MIN                       2000

Why not move this above NET_MIN (2048) ?

ANJaLI> Will move it.

**Review by hallyn — APPROVED** [review]
Best to squash the two commits, or do a squash+merge merge of it, but in any case looks good, thanks.

**Comment by anjalidk** [comment]
Hi Serge,
I see that this change has not been merged into master yet? Is there any other changes needed to be done on my part?

Anjali

**Comment by hallyn** [comment]
> Hi Serge, I see that this change has not been merged into master yet? Is there any other changes needed to be done on my part? Anjali

What do you mean?  It is merged...

**Comment by anjalidk** [comment]
If I check out a fresh copy of libresource I don't see the route QA infra or the tests/route directory added. I think I still see the pull open there. I added new commit to it today
Anjali

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by hallyn** [comment]
The 'QA infra for ROUTE' was pushed to https://github.com/lxc/libresource/pull/24 .

For that one, I'd ask you to rebase the whole thing to drop the merge, and squash the first and second commits.

**Comment by anjalidk** [comment]
> On Jul 20, 2023, at 3:41 PM, Serge Hallyn ***@***.***> wrote:
> 
> 
> The 'QA infra for ROUTE' was pushed to #24 .
> 
> For that one, I'd ask you to rebase the whole thing to drop the merge, and squash the first and second commits.

Thanks, the merge added the tests/.. directory. Let me drop the merge (in git rebase), squash 1st & 2nd commits and check..
Anjali

**Comment by anjalidk** [comment]
> On Jul 21, 2023, at 12:35 PM, Anjali Kulkarni ***@***.***> wrote:
> 
> 
> 
>> On Jul 20, 2023, at 3:41 PM, Serge Hallyn ***@***.***> wrote:
>> 
>> 
>> The 'QA infra for ROUTE' was pushed to #24 .
>> 
>> For that one, I'd ask you to rebase the whole thing to drop the merge, and squash the first and second commits.
> 
> Thanks, the merge added the tests/.. directory. Let me drop the merge (in git rebase), squash 1st & 2nd commits and check..
> Anjali 

Done, let me know if that looks ok.

**Comment by anjalidk** [comment]
Looks like there were merge conflicts - resolved these via web UI - please let me know if it looks ok now.
Anjali

**Pull request merged**